### PR TITLE
P2494R2 Relaxing range adaptors to allow for move only types

### LINKED ISSUE #5617: [2022-07 LWG Motion 30] P2494R2 Relaxing range adaptors to allow for move only types
[P2494R2](https://wg21.link/p2494r2)

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes #5617
Fixes cplusplus/papers#1156

### CODE REVIEW

File: source/ranges.tex [hunk 1]
```diff
@@ -2341,15 +2341,15 @@
 \indexlibraryglobal{single_view}%
 \begin{codeblock}
 namespace std::ranges {
-  template<@\libconcept{copy_constructible}@ T>
+  template<@\libconcept{move_constructible}@ T>
     requires is_object_v<T>
   class single_view : public view_interface<single_view<T>> {
   private:
-    @\exposidnc{copyable-box}@<T> @\exposidnc{value_}@;             // \expos{} (see \ref{range.copy.wrap})
+    @\exposidnc{movable-box}@<T> @\exposidnc{value_}@;             // \expos{} (see \ref{range.move.wrap})
```
Comment (JohelEGP) [inline source/ranges.tex:2348]:
```suggestion
    @\exposidnc{movable-box}@<T> @\exposidnc{value_}@;              // \expos{} (see \ref{range.move.wrap})
```

File: source/ranges.tex [hunk 2]
```diff
@@ -3565,10 +3566,18 @@
 
 \pnum
 \recommended
-\tcode{\exposid{copyable-box}<T>} should store only a \tcode{T}
-if either \tcode{T} models \libconcept{copyable} or
+\begin{itemize}
+\item
+If \tcode{\exposid{copy_constructible}<T>} is \tcode{true},
```
Comment (JohelEGP) [inline source/ranges.tex:3571]:
```suggestion
If \tcode{\libconcept{copy_constructible}<T>} is \tcode{true},
```

File: source/ranges.tex [hunk 3]
```diff
@@ -3841,7 +3850,7 @@
   class filter_view : public view_interface<filter_view<V, Pred>> {
   private:
     V @\exposid{base_}@ = V();                              // \expos
-    @\exposidnc{copyable-box}@<Pred> @\exposid{pred_}@;                   // \expos
+    @\exposidnc{movable-box}@<Pred> @\exposid{pred_}@;                   // \expos
```
Comment (JohelEGP) [inline source/ranges.tex:3853]:
```suggestion
    @\exposidnc{movable-box}@<Pred> @\exposid{pred_}@;                    // \expos
```

File: source/ranges.tex [hunk 4]
```diff
@@ -4274,7 +4283,7 @@
     template<bool> struct @\exposid{sentinel}@;             // \expos
 
     V @\exposid{base_}@ = V();                              // \expos
-    @\placeholdernc{copyable-box}@<F> @\exposid{fun_}@;                       // \expos
+    @\placeholdernc{movable-box}@<F> @\exposid{fun_}@;                       // \expos
```
Comment (JohelEGP) [inline source/ranges.tex:4286]:
```suggestion
    @\placeholdernc{movable-box}@<F> @\exposid{fun_}@;                        // \expos
```

File: source/ranges.tex [hunk 5]
```diff
@@ -5213,7 +5222,7 @@
     template<bool> class @\exposidnc{sentinel}@;                      // \expos
 
     V @\exposid{base_}@ = V();                                      // \expos
-    @\exposidnc{copyable-box}@<Pred> @\exposid{pred_}@;                           // \expos
+    @\exposidnc{movable-box}@<Pred> @\exposid{pred_}@;                           // \expos
```
Comment (JohelEGP) [inline source/ranges.tex:5225]:
```suggestion
    @\exposidnc{movable-box}@<Pred> @\exposid{pred_}@;                            // \expos
```

File: source/ranges.tex [hunk 6]
```diff
@@ -5564,7 +5573,7 @@
 
   private:
     V @\exposid{base_}@ = V();                                      // \expos
-    @\placeholder{copyable-box}@<Pred> @\exposid{pred_}@; @\itcorr[-1]@                          // \expos
+    @\placeholder{movable-box}@<Pred> @\exposid{pred_}@; @\itcorr[-1]@                          // \expos
```
Comment (JohelEGP) [inline source/ranges.tex:5576]:
```suggestion
    @\placeholder{movable-box}@<Pred> @\exposid{pred_}@; @\itcorr[-1]@                           // \expos
```

File: source/ranges.tex [hunk 7]
```diff
@@ -9296,12 +9305,12 @@
...
 \begin{codeblock}
 namespace std::ranges {
-  template<@\libconcept{copy_constructible}@ F, @\libconcept{input_range}@... Views>
+  template<@\libconcept{mmove_constructible}@ F, @\libconcept{input_range}@... Views>
     requires (@\libconcept{view}@<Views> && ...) && (sizeof...(Views) > 0) && is_object_v<F> &&
               @\libconcept{regular_invocable}@<F&, range_reference_t<Views>...> &&
               @\exposconcept{can-reference}@<invoke_result_t<F&, range_reference_t<Views>...>>
   class zip_transform_view : public view_interface<zip_transform_view<F, Views...>> {
-    @\exposidnc{copyable-box}@<F> @\exposid{fun_}@;                   // \expos
+    @\exposidnc{movable-box}@<F> @\exposid{fun_}@;                   // \expos
```
Comment (JohelEGP) [inline source/ranges.tex:9313]:
```suggestion
    @\exposidnc{movable-box}@<F> @\exposid{fun_}@;                    // \expos
```

File: source/ranges.tex [hunk 8]
```diff
@@ -10470,12 +10479,12 @@
...
 \begin{codeblock}
 namespace std::ranges {
-  template<@\libconcept{forward_range}@ V, @\libconcept{copy_constructible}@ F, size_t N>
+  template<@\libconcept{forward_range}@ V, @\libconcept{move_constructible}@ F, size_t N>
    requires @\libconcept{view}@<V> && (N > 0) && is_object_v<F> &&
             @\libconcept{regular_invocable}@<F&, @\exposid{REPEAT}@(range_reference_t<V>, N)...> &&
             @\exposconcept{can-reference}@<invoke_result_t<F&, @\exposid{REPEAT}@(range_reference_t<V>, N)...>>
   class adjacent_transform_view : public view_interface<adjacent_transform_view<V, F, N>> {
-    @\exposidnc{copyable-box}@<F> @\exposid{fun_}@;                       // \expos
+    @\exposidnc{movable-box}@<F> @\exposid{fun_}@;                       // \expos
```
Comment (JohelEGP) [inline source/ranges.tex:10487]:
```suggestion
    @\exposidnc{movable-box}@<F> @\exposid{fun_}@;                        // \expos
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]
Mostly missing spaces to align `// \expos`.

**Comment by tkoeppe** [comment]
Several changes to the nested iterator types of the views don't seem to be in the paper, but I take it that's an error in the paper?

**Comment by JohelEGP** [comment]
If you mean that an updated _template-head_ of a range adaptor is not reflected in the _template-head_ of the class synopsis of an inner class (template), then you're right.

File: source/ranges.tex [hunk 9]
```diff
@@ -4072,31 +4073,32 @@
 the initialization of the bound argument entities of the result,
 as specified above, are all well-formed.
 
-\rSec2[range.copy.wrap]{Copyable wrapper}
+\rSec2[range.move.wrap]{Copyable wrapper}
```
Comment (JohelEGP) [inline source/ranges.tex:4076]: As requested by the paper.
```suggestion
\rSec2[range.move.wrap]{Movable wrapper}
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

**Comment by JohelEGP** [comment]
> If you mean that an updated _template-head_ of a range adaptor is not reflected in the _template-head_ of the class synopsis of an inner class (template), then you're right.

But the changes in this PR already takes account for those.

**Review by JohelEGP — CHANGES_REQUESTED** [review]
The paper fails to account for the use of _`copyable-box`_ in [range.chunk.by.view]. I believe simply changing its use to _`movable-box`_, as done with `filter_view`, will suffice.

**Comment by tkoeppe** [comment]
> > If you mean that an updated _template-head_ of a range adaptor is not reflected in the _template-head_ of the class synopsis of an inner class (template), then you're right.
> 
> But the changes in this PR already takes account for those.

Right, I saw that, and wanted to make sure that was deliberate. I'll mention this in the commit message. Thanks!

File: source/ranges.tex [hunk 10]
```diff
@@ -3365,21 +3365,22 @@
...
-    @\exposid{copyable-box}@<W> @\exposid{value_}@ = W();   // \expos, see \ref{range.copy.wrap}
+    @\exposid{movable-box}@<W> @\exposid{value_}@ = W();    // \expos, see \ref{range.move.wrap}
     Bound @\exposid{bound_}@ = Bound();    // \expos
 
   public:
     repeat_view() requires @\libconcept{default_initializable}@<W> = default;
 
-    constexpr explicit repeat_view(const W& value, Bound bound = Bound());
+    constexpr explicit repeat_view(const W& value, Bound bound = Bound())
+      requires @\libconcept{copy_constructible}@<W>;
```
Comment (JohelEGP) [inline source/ranges.tex:3383]: The corresponding change is missing in its detailed specifications.

File: source/ranges.tex [hunk 11]
```diff
@@ -13302,7 +13302,7 @@
     requires @\libconcept{view}@<V> && is_object_v<Pred>
   class chunk_by_view : public view_interface<chunk_by_view<V, Pred>> {
     V @\exposid{base_}@ = V();                       // \expos
-    @\exposid{copyable-box}@<Pred> @\exposid{pred_}@ = Pred();  // \expos
+    @\exposid{movable-box}@<Pred> @\exposid{pred_}@ = Pred();  // \expos
```
Comment (JohelEGP) [inline source/ranges.tex:13306]: A space is needed before the comment to keep them aligned.

File: source/ranges.tex (hunk 10 above)
Comment (tkoeppe) [reply source/ranges.tex:3383]: Thanks, I'll fix it later

File: source/ranges.tex (hunk 11 above)
Comment (tkoeppe) [reply source/ranges.tex:13306]: Will fix!

**Pull request merged**